### PR TITLE
[HIVEMALL-78] Implement AUC UDAF for binary classification

### PR DESCRIPTION
**Pull request by takuti**
## What changes were proposed in this pull request?

In addition to current `auc(array, array)` for ranking (myui/hivemall#326), this patch supports `auc(double, double)` for binary classification.

## What type of PR is it?

Feature

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-78

## How was this patch tested?

Created unit test for the UDAF, and passed:

```
$ mvn -Dtest=hivemall.evaluation.AUCUDAFTest test
```

Moreover, I have launched manual tests by the following queries:

```sql
with data as (
  select 0.5 as prob, 0 as label
  union all
  select 0.3 as prob, 1 as label
  union all
  select 0.2 as prob, 0 as label
  union all
  select 0.8 as prob, 1 as label
  union all
  select 0.7 as prob, 1 as label
), data_ordered as (
  select prob, label
  from data
  order by prob desc
)
select auc(prob, label)
from (
  select prob, label
  from data_ordered
  distribute by floor(prob / 0.2)
) t;
```

```sql
with data as (
  select 0.5 as prob, 0 as label
  union all
  select 0.3 as prob, 1 as label
  union all
  select 0.2 as prob, 0 as label
  union all
  select 0.8 as prob, 1 as label
  union all
  select 0.7 as prob, 1 as label
), data_ordered as (
  select prob, label
  from data
  order by prob desc
)
select auc(prob, label)
from data_ordered;
```

Both showed `AUC=0.83333`. This result is same as [scikit-learn's roc_auc_score()](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html):

```
>>> roc_auc_score([0,1,0,1,1],[0.5,0.3,0.2,0.8,0.7])
0.83333333333333326
```

## How to use this feature?

See above queries. Input data needs to be ordered by scores in a descending order.

### CODE REVIEW

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 1]
```diff
@@ -49,35 +50,251 @@
...
+            ClassificationAUCAggregationBuffer myAggr = (ClassificationAUCAggregationBuffer) agg;
+            myAggr.reset();
+        }
+
+        @Override
+        public void iterate(AggregationBuffer agg, Object[] parameters) throws HiveException {
+            ClassificationAUCAggregationBuffer myAggr = (ClassificationAUCAggregationBuffer) agg;
+
+            double score = HiveUtils.getDouble(parameters[0], scoreOI);
+            double label = HiveUtils.getDouble(parameters[1], labelOI);
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]:
```java
if(parameters[0] == null) {
   continue;
}
if(parameters[1]) == null) { // separate ^ for debugging (e.g., attaching breakpoints)
   continue;
}

int label = PrimitiveObjectInspectorUtils.getInt(parameters[1], labelOI);
if(label == -1) {
   label = 0;
} else if (label != 0 && label != 1) {
   throw new UDFArgumentException("label MUST be 0/1 or -1/1: " + label);
}

```
https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/PrimitiveObjectInspectorUtils.java#L634

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 2]
```diff
@@ -49,35 +50,251 @@
...
+
+        @Override
+        public ObjectInspector init(Mode mode, ObjectInspector[] parameters) throws HiveException {
+            assert (parameters.length == 2 || parameters.length == 3) : parameters.length;
+            super.init(mode, parameters);
+
+            // initialize input
+            if (mode == Mode.PARTIAL1 || mode == Mode.COMPLETE) {// from original data
+                this.scoreOI = (PrimitiveObjectInspector) parameters[0];
+                this.labelOI = (PrimitiveObjectInspector) parameters[1];
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]:
```java
this.scoreOI = HiveUtils.asDoubleCompatibleOI(parameters[0]);
this.labelOI = HiveUtils.asIntegerOI(parameters[1]);
```

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 3]
```diff
@@ -49,35 +50,251 @@
...
         if (typeInfo.length != 2 && typeInfo.length != 3) {
             throw new UDFArgumentTypeException(typeInfo.length - 1,
                 "_FUNC_ takes two or three arguments");
         }
 
-        ListTypeInfo arg1type = HiveUtils.asListTypeInfo(typeInfo[0]);
-        if (!HiveUtils.isPrimitiveTypeInfo(arg1type.getListElementTypeInfo())) {
-            throw new UDFArgumentTypeException(0,
-                "The first argument `array rankItems` is invalid form: " + typeInfo[0]);
+        if (HiveUtils.isNumberTypeInfo(typeInfo[0]) && HiveUtils.isNumberTypeInfo(typeInfo[1])) {
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: `&& HiveUtils.isIntegerTypeInfo(typeInfo[1])`

Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: `&& HiveUtils.isIntegerTypeInfo(typeInfo[1])`

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java (hunk 1 above)
Comment (takuti) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: `continue;` -> `return;`

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 4]
```diff
@@ -49,35 +51,264 @@
...
+
+        double get() {
+            if (tp == 0 || fp == 0) {
+                return 0.d;
+            }
+            double res = a + trapezoidArea(fp, fpPrev, tp, tpPrev);
+            return res / (tp * fp); // scale
+        }
+
+        void iterate(@Nonnull double score, @Nonnull int label) {
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: Remove `@Nonnull` for primitives (never become null)

File: docs/gitbook/eval/auc.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  KIND, either express or implied.  See the License for the
+  specific language governing permissions and limitations
+  under the License.
+-->
+
+<!-- toc -->
+
+# Area Under the ROC Curve
+
+[ROC curve](https://en.wikipedia.org/wiki/Receiver_operating_characteristic) and Area Under the ROC Curve (AUC) are widely-used metric for binary (i.e., positive or negative) classification problems such as [Logistic Regression](../binaryclass/a9a_lr.md).
```
Comment (myui) [inline docs/gitbook/eval/auc.md]: fix the link to `../binaryclass/a9a_lr.html`

File: docs/gitbook/eval/auc.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  order by prob desc
+)
+select auc(prob, label)
+from (
+  select prob, label
+  from data_ordered
+  distribute by floor(prob / 0.2)
+) t;
+```
+
```
Comment (myui) [inline docs/gitbook/eval/auc.md]: Add a note explaining what `floor(prob / 0.2)` is meaning. Distribute AUC computation into 5 bins.

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 7]
```diff
@@ -49,35 +51,264 @@
...
+            ClassificationAUCAggregationBuffer myAggr = (ClassificationAUCAggregationBuffer) agg;
+
+            if (parameters[0] == null) {
+                return;
+            }
+            if (parameters[1] == null) {
+                return;
+            }
+
+            double score = HiveUtils.getDouble(parameters[0], scoreOI);
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: if(score < 0.0d || score > 1.0d) {
   throw new UDFArgumentException("score value MUST be in range [0,1]: " + score);
}

**Comment by myui** [comment]
LGTM. Merged. Thank you so much for a big contribution!

**Pull request closed without merging**